ships/0027-image-abstraction-layer.md

```diff
@@ -0,0 +1,232 @@
+
+Once such a custom_ resource_ is created (with `kubectl create`, for example) the operator won’t
+act on it immediately but it will instead wait until an ImageImport, referring to the created
+Image object, is instantiated by the user.
+
+An ImageImport object looks like:
+
+```
+apiVersion: shipwright.io/v1alpha1
+kind: ImageImport
+metadata:
+  name: myapp-import-00
+  namespace: my-namespace
+Spec:
+  targetImage: myapp-devel
+  from: quay.io/company/myapp:latest
+  mirror: false
+  insecure: false
+```
+
+All Image properties also show up in an ImageImport object, so, if the user does not provide data
+for these properties in the ImageImport the operator will use whatever is set in the Image instead.
+Follow a list of meaning for each property in an ImageImport.Spec:
+
+| Property    | Description  |
+| ----------- | ------------ |
+| targetImage | The Image to which this ImageImport refers to. This is a mandatory field. (required) |
+| from        | Indicates the source of the image (from where the operator should import it), if not set the operator will use the spec.targetImage.From by default. (optional) |
+| mirror      | Informs if the Image should be mirrored in the configured mirror registry. If not set targetImage.Mirror is used instead. (optional) |
+| insecure    | Indicates if the remote registry hosting the image is using “invalid certificates”, i.e. skip TLS verification. If not set spec.targetImage.Insecure is used instead. (optional) |
+
+If an ImageImport has in its `spec.targetImage` an Image that does not exist yet one should be
+created by the operator before starting to process the ImageImport object.
+
+### Mirroring images
+
+If mirroring is set in an Image or ImageImport the operator will attempt to mirror the image
+content into another registry provided by the user. To mirror images locally one needs to inform
+the operator about the mirror registry location and authentication methods. Such mirror registry
+could be configured through a Secret called `mirror-registry-config`, this secret would contain
+the following properties:
+
+| Name       | Description                                                                       |
+| -----------| --------------------------------------------------------------------------------- |
+| address    | The mirror registry URL                                                           |
+| username   | Username                                                                          |
+| password   | The password to be used                                                           |
+| token      | The auth token to be used                                                         |
+| insecure   | Allows access to insecure registries if set to "true"                             |
+| repository | If set all images are going to be mirrored inside this registry repository        |
+
+### Image status
```

Comment:
Assuming that the operator constantly reacts on a single `ImageImport` which in consequence reference an `Image`(just stating the obvious), why do we need to persist a state on the `Image` crd, as this just serves as a configuration object. Why wouldn't we opt-in for this being part of the `ImageImport` status?

Context: We tried to have this distinction in Build vs BuildRun, we still persist a state in the Build, but as this is only a configuration object, we will drop that at some point. Only the BuildRun should persist a state.

Reply:
I think that to have everything gathered inside the Image object brings us two things:

1. It becomes simpler for other operators to observe just one object (Image) instead of multiple objects (ImageImports) for the same Image.
2. We can discard already processed ImageImports, lowering the number of entities in the storage.

Reply:
ok, as long as we always have this `1-to-N` relationship, that's fine.